### PR TITLE
Update oneDNN specification to 0.8

### PR DESCRIPTION
**Pull request by rsdubtso**
- More detailed description of execution model
- Skeleton of the data model
- Skeleton of the primitives descriptions

### CODE REVIEW

File: source/_static/custom.js [hunk 1]
```diff
@@ -13,3 +13,41 @@ var wapSection = "oneapi";
...
 })(); 
+
+window.MathJax = {
+    TeX: {
+	Macros: {
+	    src: '\\operatorname{src}',
+	    srclayer: '\\operatorname{src\\_layer}',
+	    srciter: '\\operatorname{src\\_iter}',
+	    srciterc: '\\operatorname{src\\_iter\\_c}',
+	    weights: '\\operatorname{weights}',
```
Comment (kolbusa) [inline source/_static/custom.js:24]: Test.

File: source/elements/oneDNN/source/introduction.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+.. _`open source implementation`: https://github.com/oneapi-src/oneDNN/tree/dev-v2
+
+Introduction
+------------
+
+Although the origins of this specification are in the existing `open source
+implementation`_, its goal is to define a *portable* set of APIs.  To this
+end, for example, it intentionally omits implementation-specific details like
+tiled or blocked memory formats (layouts), and instead describes plain
+multi-dimensional memory formats and defines opaque _optimized_ memory format
```
Comment (emfomenk) [inline source/elements/oneDNN/source/introduction.rst]: don't know why, `_optimized_` isn't rendered correctly

File: source/elements/oneDNN/source/introduction.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+sometimes represent more complex *fused* computations such as a forward
+convolution followed by a ReLU. Fusion, among other things, is controlled via
+the *primitive attrubutes* mechanism.
+
+The most important difference between a primitive and a pure function is that
+a primitive can be specialized for a subset of input parameters.
+
+For example, a convolution primitive stores parameters like tensor shapes and
+can pre-compute other dependent parameters like cache blocking. This approach
+allows DNNL primitives to pre-generate code specifically tailored for the
```
Comment (emfomenk) [inline source/elements/oneDNN/source/introduction.rst]: DNNL or oneDNN?

File: source/elements/oneDNN/source/execution_model/index.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+Execution Model
+---------------
+
+To execute a primitive, a user needs to pass memory arguments and a stream to
+the :any:`dnnl::primitive::execute` member function.
+
+.. image:: ../_static/img_execution_model.png
+   :width: 600
+   :alt: oneDNN execution model
```
Comment (emfomenk) [inline source/elements/oneDNN/source/execution_model/index.rst:14]: what does arrow from stream to engine mean? IMO, picture is a little bit confusing...

File: source/elements/oneDNN/source/data_model/data_types/index.rst [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+   :project: oneDNN
+
+.. namespace:: dnnl::memory::data_type
+
+oneDNN supports training and inference with the following data types:
+
+========== ===========================================
+Usage mode Data types
+========== ===========================================
+Inference  :any:`f32`, :any:`bf16`, :any:`s8`/:any:`u8`
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/data_types/index.rst]: f16?

File: source/elements/oneDNN/source/data_model/memory/formats.rst [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+           strides_transposed};
+
+Format Tags
+~~~~~~~~~~~
+
+In addition to strides, oneDNN provides named *format tags* via the
+:any:`dnnl::memory::format_tag` enum type. The enumerators of this type can be
+used instead of strides for dense plain layouts.
+
+The format tag names for _N_-dimensional memory formats use first _N_ letters
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/memory/formats.rst]: same: `_N_` isn't rendered.

File: source/elements/oneDNN/source/data_model/memory/formats.rst [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+More precisely, let :math:`T` be a tensor of rank :math:`n` and let
+:math:`\sigma` be the permutation of the |strides| array that sorts it, i.e.
+:math:`\operatorname{strides}[\sigma(i)] \leq
+\operatorname{strides}[\sigma(j)]` if :math:`\sigma(i) < \sigma(j)` for all
+:math:`0 \leq i, j < n`. Then the following must hold:
+
+.. math::
+   \operatorname{stides}[j] \geq
+   \operatorname{strides}[j + 1] * \operatorname{dimensions}[\sigma(j)]
+   \text{ for all } 0 \leq j < n - 1.
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/memory/formats.rst]: I am not sure I follow... I would expect `stride[j]` be always multiplied by `dim[j]`. However, the relation could invoke sigma. Something like:
```
Let sigma(i) be such that:
    sigma(i) > sigma(j) => stride[i] <= stride[j]
then
    stride[i] * dim[i] <= stride[j] for all i,j : sigma(i) > sigma(j)
```

File: source/elements/oneDNN/source/data_model/memory/formats.rst [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+not possible to omit |H| when |D| is present and it is never possible to omit
+|W|. Canonical interpretation is documented for each primitive. However, this
+means that the |strides| array plays an important role defining the order in
+which different dimension are laid out in memory. Moreover, the |strides| need
+to agree with |dimensions|.
+
+More precisely, let :math:`T` be a tensor of rank :math:`n` and let
+:math:`\sigma` be the permutation of the |strides| array that sorts it, i.e.
+:math:`\operatorname{strides}[\sigma(i)] \leq
+\operatorname{strides}[\sigma(j)]` if :math:`\sigma(i) < \sigma(j)` for all
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/memory/formats.rst]:
```suggestion
:math:`\operatorname{strides}[i] \leq
\operatorname{strides}[j]` if :math:`\sigma(i) > \sigma(j)` for all
```

File: source/elements/oneDNN/source/execution_model/index.rst (hunk 4 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/execution_model/index.rst:14]: It is. I'll rework it, but I'm keeping it as a placeholder.

File: source/elements/oneDNN/source/data_model/memory/formats.rst (hunk 7 above)
Comment (emfomenk) [reply source/elements/oneDNN/source/data_model/memory/formats.rst]: >  I would expect stride[j] be always multiplied by dim[j]. 

Just to make clear this statement clear(er): I would expect the same indices for stride and dim, while in the text strides use j, but dims use sigma(j)

Comment (rsdubtso) [reply source/elements/oneDNN/source/data_model/memory/formats.rst]: Yeah. You are right. The original text assumes that the strides are sorted and they are not. Not sure I came around to writing it this way :) I've fixed it but changed the notation a bit. Please take a look.

File: source/elements/oneDNN/source/data_model/memory/formats.rst (hunk 8 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/data_model/memory/formats.rst]: Fixed.

File: source/elements/oneDNN/source/data_model/memory/formats.rst (hunk 7 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/data_model/memory/formats.rst]: I've fixed this but changed the notation a bit.

Comment (emfomenk) [reply source/elements/oneDNN/source/data_model/memory/formats.rst]: LGTM, thanks

File: source/elements/oneDNN/source/data_model/memory/formats.rst [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+More precisely, let :math:`T` be a tensor of rank :math:`n` and let
+:math:`\sigma` be the permutation of the |strides| array that sorts it, i.e.
+:math:`\operatorname{strides}[i] \geq
+\operatorname{strides}[j]` if :math:`\sigma(i) < \sigma(j)` for all
+:math:`0 \leq i, j < n`. Then the following must hold:
+
+.. math::
+   \operatorname{stides}[i] \geq
+   \operatorname{strides}[j] * \operatorname{dimensions}[j]
+   \text{ if } \sigma(i) < \sigma(j) \text{ for all } 0 \leq i, j < n - 1.
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/memory/formats.rst]: n-1 --> n

File: source/elements/oneDNN/source/primitives/index.rst [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+complex fused computations such as a forward convolution followed by a ReLU.
+
+The most important difference between a primitive and a pure function is
+that a primitive can store state.
+
+One part of the primitive's state is immutable. For example, convolution
+primitives store parameters like tensor shapes and can pre-compute other
+dependent parameters like cache blocking. This approach allows DNNL primitives
+to pre-generate code specifically tailored for the operation to be performed.
+The DNNL programming model assumes that the time it takes to perform the
```
Comment (emfomenk) [inline source/elements/oneDNN/source/primitives/index.rst]: 2 times: DNNL --> oneDNN

(let me know if you intentionally preserved DNNL here)

File: source/elements/oneDNN/source/introduction.rst [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+| Implementation           | Memory object     | Primitive            |
++--------------------------+-------------------+----------------------+
+
+.. rubric:: General API notes
+
+There are certain assumptions on how oneDNN objects behave:
+
+- Memory and operation descriptors behave similarly to trivial types. Their
+  default constructors (only available for memory descriptors) create *zero*
+  objects that indicate absence of a memory object.  Copying memory and
```
Comment (emfomenk) [inline source/elements/oneDNN/source/introduction.rst]: IMO, weird phrase: "their (mem and op) default ctors ... that indicate absence of a **memory** object", i.e. we started with op and mem, but finished with mems only.

Not even sure what you wanted to highlight here: the expensiveness of copying the objects, or the availability and behavior of _default_ constructors? Is it important?

File: source/elements/oneDNN/source/introduction.rst [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+All oneDNN functions throw the following exception in case of error.
+
+.. doxygenstruct:: dnnl::error
+   :project: oneDNN
+   :members:
+
+Additionally, many oneDNN functions that construct or return oneDNN objects
+have a boolean ``allow_empty`` parameter that defaults to ``false`` and that
+makes the library to return an empty object when an object cannot be
```
Comment (emfomenk) [inline source/elements/oneDNN/source/introduction.rst]:
```suggestion
makes the library to return a zero objec (for memory descriptors) or an empty object
(for other objects) when an object cannot be
```

**Review by emfomenk — APPROVED** [review]
LGTM

File: source/elements/oneDNN/source/primitives/index.rst (hunk 10 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/primitives/index.rst]: No, this is not intentional. I thought I fixed these :/

File: source/elements/oneDNN/source/introduction.rst (hunk 11 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/introduction.rst]: Reworded (deleted :))

File: source/elements/oneDNN/source/introduction.rst (hunk 12 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/introduction.rst]: Fixed.

File: source/elements/oneDNN/source/data_model/data_types/bfloat16.rst [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+IEEE 16-bit half-precision floating point data type, :any:`f16`.  Compared
+directly to :any:`f16`, which has a 5-bit exponent and a 10-bit mantissa,
+:any:`bf16` trades increased range for reduced precision.
+
+.. image:: ../../_static/img_bf16_diagram.png
+   :width: 600
+   :alt: Diagram depicting the bit-wise layout of f32, bf16, and f16 floating point data types
+
+More details of the bfloat16 data type can be found `here
+<https://en.wikipedia.org/wiki/Bfloat16_floating-point_format>`__.
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/data_types/bfloat16.rst:27]: What is __ here?

**Review by emfomenk — CHANGES_REQUESTED** [review]

File: source/elements/oneDNN/source/data_model/data_types/bfloat16.rst (hunk 13 above)
Comment (rsdubtso) [reply source/elements/oneDNN/source/data_model/data_types/bfloat16.rst:27]: It's just a syntax for links.

File: source/elements/oneDNN/source/data_model/memory/formats.rst [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+   \text{ if } \sigma(i) < \sigma(j) \text{ for all } 0 \leq i, j < n.
+
+For an element with coordinates :math:`(i_0, \ldots, i_{n-1})` such that
+:math:`0 \leq i_j < \operatorname{dimensions}[j]` for :math:`0 \leq j < n`, its
+offset in memory is computed as:
+
+.. math::
+
+   \operatorname{offset}(i_0, \ldots, i_{n-1})
+   = \sum_{j=0}^{n-1} i_j * \operatorname{strides}[j].
```
Comment (emfomenk) [inline source/elements/oneDNN/source/data_model/memory/formats.rst:77]: btw, do we want to mention `offset0`?

Comment (rsdubtso) [reply source/elements/oneDNN/source/data_model/memory/formats.rst:77]: Makes sense, but not now. Putting this on my TODO list.

**Pull request merged**